### ISSUE TITLE
sdk-connectivity - Bug with Cache-Handling results in 401 due to logic-error

### ISSUE THREAD

**Issue by drvup** [SAP/cloud-sdk-js#4064]
Dear SAP SDK Team, 

we've updated some applications to the newest version of the SAP Cloud SDK "3.4.0" and identified issues within the destination-handling. After some time, the connections result in a http-status 401 without any sense, as they did work fine before. The code basis on the apps wasn't touched on any relevant place. 

**Describe the bug**
After some time (I guess, after 12 hours, because that's the `expiresIn` value of the JWT token, fetched before via destination-service), the connections to destinations do fail, if we're using the `cache = true` option.

**Expected behavior**
If the cache flag is set to `true`, we do expect the cache handling for destinations is working and not valid JWTs are getting invalid / deleted from the App-Cache.

**Possible Root Causes**
As I was pretty sure it's not about our own logic, I did check the [Release Notes](https://sap.github.io/cloud-sdk/docs/js/release-notes) and I identified some changes by @marikaner and others on the related part. Destination and Destination-Cache-Handling. So I checked the changes and did debug the boilerplate code a bit. 
Found the possible root cause for this issue on file `destination-from-vcap.js` at method `getDestinationFromServiceBinding()`. 
Here, the destination is fetched from the cache (in our case, it is):
```
   const destination = (await retrieveDestinationFromCache(retrievalOptions)) ||
        (await retrieveDestinationWithoutCache(retrievalOptions));
```

But unfortunately, the already cached destination is saved again to the cache some lines below:
```
    if (options.useCache) {
        // As the grant type is clientCredential, isolation strategy is 'tenant'.
        await destination_cache_1.destinationCache.cacheRetrievedDestination((0, jwt_1.decodeOrMakeJwt)(options.jwt), destWithProxy, 'tenant');
    }
```

This results in a new cache entry with a expiresIn-Timestamp (set by the SDK) based on the `expiresIn` values of the already cached JWT. So this `expiresIn` value is already withdrawn and can't be used to set the "expiresIn-Timestamp" for the new cache entry.

This fits exactly to the scenario we're having right now and results in 401 http status errors, as soon as the bearer / JWT token is invalid.

**Screenshots**
This screenshot shows the cached bearer-token / JWT, as part of the cached-destination. It got copied and decoded at 17.08.23 09:45h UTC +2. You can see, the `exp` value is already in the past and this bearer-token is invalid already. 
![jwtDecoded](https://github.com/SAP/cloud-sdk-js/assets/9018930/81f1642a-57c4-4386-bbf5-33e534c36091)

The other screenshot shows the `expiresIn` value of the cached destination who is used to calculate the new "expiresIn-Timestamp" for the new cache entry:
![newCacheEntry](https://github.com/SAP/cloud-sdk-js/assets/9018930/82b45e50-f30f-4960-8212-57391b0ae4ae)

**Used Versions:**
- node version 18.15.0
- npm version 9.5.0
- SAP Cloud SDK version 3.4.0

**Impact / Priority**
No priority for us. We will roll back to 3.1.1 for the applications we did update to 3.4.0 - But I am really sure this is a big issue for all the SDK developers ;-)

Kind Regards,
Cedric

**Comment by deekshas8** [comment]
Hi @drvup ,

Good catch :) 
This is indeed a bug in the caching implementation in `getDestinationFromServiceBinding` method. We will fix this soon and update you.

**Comment by deekshas8** [comment]
Hi @drvup ,

We fixed the issue in the latest `v3.5.0` release. Please try it out.

**Comment by jjtang1985** [comment]
I'll close this ticket as it's solved.
